### ISSUE TITLE
How to train my model on multiple GPUs?

### ISSUE THREAD

**Issue by zbw0329** [eladrich/pixel2style2pixel#154]
Hi,I am very interesting in your pspGAN,.I have two 1080Ti,I want to know how to change the code to use them both.
I tried to change the code of 'coach.py'
from   self.net = pSp(self.opts).to(self.device)
to  self.net = (torch.nn.DataParallel(pSp(self.opts)).module).to.(self.device)
But it doesn't work,so i want to get some help from you
Thanks!

**Comment by zbw0329** [comment]
And I wonder what kind of devices you use
Thanks!

**Comment by yuval-alaluf** [comment]
In our experiments, we used a single P40 gpu. 
I don't have too much experience in using `DataParallel`. However, with some adjustments pSp should be able to run on multiple GPUs. 
Did you remove the line `self.device = 'cuda:0'` at the beginning of `coach.py`?

**Comment by zbw0329** [comment]
I didn't remove the code of 'self.device',I don't know what code should be used instead.

**Comment by yuval-alaluf** [comment]
For working with multiple GPUs, I recommend exploring using Distributed Data Parallel with Pytorch. There are many great tutorials online. 
At the moment, we do not plan on extending support for multi-gpu training of pSp.

**Comment by echen01** [comment]
I made the following changes in `coach.py`, which seem to be working well: 

`self.device = "cuda"`
`self.net.encoder = nn.DataParallel(self.net.encoder)`
`self.net.decoder = nn.DataParallel(self.net.decoder)`

I also made the losses data parallel:
`self.lpips_loss.net = nn.DataParallel(self.lpips_loss.net)`
`self.id_loss.facenet = nn.DataParallel(self.id_loss.facenet)`
`self.moco_loss.model = nn.DataParallel(self.moco_loss.model)`

This thread had some helpful tips too: https://discuss.pytorch.org/t/how-to-best-use-dataparallel-with-multiple-models/39289. 

Hope this helps!

**Comment by zbw0329** [comment]
> For working with multiple GPUs, I recommend exploring using Distributed Data Parallel with Pytorch. There are many great tutorials online.
> At the moment, we do not plan on extending support for multi-gpu training of pSp.

Thanks a lot,I will have a try with that~

**Comment by zbw0329** [comment]
> I made the following changes in `coach.py`, which seem to be working well:
> 
> `self.device = "cuda"`
> `self.net.encoder = nn.DataParallel(self.net.encoder)`
> `self.net.decoder = nn.DataParallel(self.net.decoder)`
> 
> I also made the losses data parallel:
> `self.lpips_loss.net = nn.DataParallel(self.lpips_loss.net)`
> `self.id_loss.facenet = nn.DataParallel(self.id_loss.facenet)`
> `self.moco_loss.model = nn.DataParallel(self.moco_loss.model)`
> 
> This thread had some helpful tips too: https://discuss.pytorch.org/t/how-to-best-use-dataparallel-with-multiple-models/39289.
> 
> Hope this helps!

Thanks for the help,I will try this code in my model~

**Comment by LiUzHiAn** [comment]
Hi, may I know the rough training time with a single P40 GPU? for example, the seg-to-face task. And what's the exact iteration number (which is set to 500,000 by default)?

**Comment by yuval-alaluf** [comment]
We trained the segmentation to image task for 80 000 steps (but if i recall correctly, it converged much faster, e.g. after 30,000). 
This would probably take 2-3 days.

**Comment by LiUzHiAn** [comment]
Got it. Thanks.

**Comment by echen01** [comment]
It seems like you're using a very small batch size. Maybe that explains the speed issues. 

When training the model on multiple GPUs, I increased the batch size appropriately to use all available memory. 

For reference, I used a batch size of 14 for 1 GTX Titan X and 28 for 2 GPUs.

**Comment by June01** [comment]
> 

Alright. If I understand correctly, the memory size of your titan x is 12G, which is just a bit bigger than me, i.e., 10G of 3080 (a typo in the previous comment).  However, the maximum batch_size of my experiment is 2 for 1 gpu, I am sure what might result in such a big gap in batch_size between us, from 2 to 14? For reference, my training script is:
```
python scripts/train.py --dataset_type=xxx \
--exp_dir=xxx \
--workers=8 \
--batch_size=2 \
--test_batch_size=2 \
--test_workers=2\
--val_interval=2500 \
--save_interval=5000 \
--encoder_type=GradualStyleEncoder \
--start_from_latent_avg \
--lpips_lambda=0.2 \
--l2_lambda=1 \
--moco_lambda=0.5 \
--input_nc=3 \
--output_size=256
```

**Comment by diamond0910** [comment]
pSp has BatchNorm2 layer. Do multiple GPUs need to change it to SyncBatchNorm？
